### PR TITLE
Feat/popularity implementation

### PR DESCRIPTION
**Pull request by dalyathan**
We have a working code that calculates the popularityScore of products while also minimzing the number of fields queried to get that. But the setProductPopularity method of SortService saves the popularity score multiple times for a single product and the collection's popularityScore only takes into consideration the collection's products. These will be fixed in the next PR

### CODE REVIEW

File: packages/vendure-plugin-sort-by-popularity/src/sort-by-popularity.plugin.ts [hunk 1]
```diff
@@ -13,6 +15,18 @@ import { SortService } from './sort.service';
     `,
   },
   configuration: (config) => {
+    config.customFields.Product.push({
+      name: 'popularityScore',
+      type: 'float',
```
Comment (martijnvdbrug) [inline packages/vendure-plugin-sort-by-popularity/src/sort-by-popularity.plugin.ts:20]: I think we can make the popularity score an int? Let's avoid floats if they are not really needed. We can round the scores to INTs, thats fine.

File: packages/vendure-plugin-sort-by-popularity/src/popularity.controller.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { Controller, Get } from '@nestjs/common';
+import { Ctx, RequestContext } from '@vendure/core';
+import { Success } from '../../test/src/generated/admin-graphql';
+import { SortService } from './sort.service';
+@Controller('/order-by-popularity')
+export class OrderByPopularityController {
+  constructor(private sortService: SortService) {}
+  @Get('calculate-scores')
+  async findAll(@Ctx() ctx: RequestContext): Promise<Success> {
```
Comment (martijnvdbrug) [inline packages/vendure-plugin-sort-by-popularity/src/popularity.controller.ts:9]: `findAll` is a typo I think?

File: packages/vendure-plugin-sort-by-popularity/src/sort.service.ts [hunk 3]
```diff
@@ -1,12 +1,83 @@
...
-    // Get orders from pas X months
-    // Combine variant quantities per product
-    // Normalize by total quantity and store as popularity score on product
-    // Get all products and products of child collections for each collection
-    // Use sum of products per collection as popularity score
+  constructor(
+    private connection: TransactionalConnection,
+    private collectionService: CollectionService
+  ) {}
+  async setProductPopularity(ctx: RequestContext): Promise<Success> {
```
Comment (martijnvdbrug) [inline packages/vendure-plugin-sort-by-popularity/src/sort.service.ts:18]: We can do this in another PR, but:
We should make this endpoint push a job to the worker. There is an example here: https://github.com/Pinelab-studio/pinelab-vendure-plugins/blob/f45644e3f39bd7eef948e930b91dde5f823b703a/packages/vendure-plugin-sendcloud/src/api/sendcloud.service.ts#L83

File: packages/vendure-plugin-sort-by-popularity/src/sort.service.ts [hunk 4]
```diff
@@ -1,12 +1,83 @@
...
+      .addSelect(['product.deletedAt', 'product.enabled', 'product.id'])
+      .innerJoin('productVariant.collections', 'collection')
+      .addSelect(['collection.id'])
+      .innerJoin('`order`.channels', 'order_channel')
+      .andWhere('`order`.orderPlacedAt is NOT NULL')
+      .andWhere('product.deletedAt IS NULL')
+      .andWhere('productVariant.deletedAt IS NULL')
+      .andWhere('product.enabled')
+      .andWhere('productVariant.enabled')
+      .andWhere(`order_channel.id=${ctx.channelId}`)
```
Comment (martijnvdbrug) [inline packages/vendure-plugin-sort-by-popularity/src/sort.service.ts:44]: Please always use `.andWhere('order_channel.id = :id', { id: ctx.channelId })` to prevent SQL injection

File: packages/vendure-plugin-sort-by-popularity/src/popularity.controller.ts (hunk 2 above)
Comment (dalyathan) [reply packages/vendure-plugin-sort-by-popularity/src/popularity.controller.ts:9]: Yes I will rename it

File: packages/vendure-plugin-sort-by-popularity/src/sort-by-popularity.plugin.ts (hunk 1 above)
Comment (dalyathan) [reply packages/vendure-plugin-sort-by-popularity/src/sort-by-popularity.plugin.ts:20]: sure I will change it to int

File: packages/vendure-plugin-sort-by-popularity/src/sort.service.ts (hunk 3 above)
Comment (dalyathan) [reply packages/vendure-plugin-sort-by-popularity/src/sort.service.ts:18]: Ok I will use the worker for that

File: packages/vendure-plugin-sort-by-popularity/src/sort.service.ts (hunk 4 above)
Comment (dalyathan) [reply packages/vendure-plugin-sort-by-popularity/src/sort.service.ts:44]: ok

File: packages/vendure-plugin-sort-by-popularity/src/sort.service.ts [hunk 5]
```diff
@@ -1,12 +1,83 @@
...
+        'orderLine.productVariant',
+        'orderLine.order',
+      ])
+      .innerJoin('orderLine.productVariant', 'productVariant')
+      .addSelect([
+        'productVariant.deletedAt',
+        'productVariant.enabled',
+        'productVariant.id',
+      ])
+      .innerJoin('orderLine.order', '`order`')
```
Comment (martijnvdbrug) [inline packages/vendure-plugin-sort-by-popularity/src/sort.service.ts:33]: Do we need the backticks (\`) here? It would be cleaner to just use `'order'`

File: packages/vendure-plugin-sort-by-popularity/src/sort.service.ts [hunk 6]
```diff
@@ -1,12 +1,83 @@
...
+      .andWhere('productVariant.deletedAt IS NULL')
+      .andWhere('product.enabled')
+      .andWhere('productVariant.enabled')
+      .andWhere(`order_channel.id=${ctx.channelId}`)
+      .addGroupBy('product.id')
+      .addOrderBy('count', 'DESC')
+      .getRawMany();
+    const maxCount = groupedOrderLines[0].count;
+    const maxValue = 1000;
+    const productRepositoty =
```
Comment (martijnvdbrug) [inline packages/vendure-plugin-sort-by-popularity/src/sort.service.ts:50]: Typo `productRepositoty`

File: packages/vendure-plugin-sort-by-popularity/src/sort.service.ts [hunk 7]
```diff
@@ -1,12 +1,83 @@
...
+    const uniqueCollectioIds: string[] = [];
+    const collectionScoreValues: number[] = [];
+    for (const groupLines of groupedOrderLines) {
+      const score = (groupLines.count * maxValue) / maxCount;
+      await productRepositoty.update(groupLines.product_id, {
+        customFields: {
+          popularityScore: score,
+        },
+      });
+      const collectionIndex = uniqueCollectioIds.findIndex(
```
Comment (martijnvdbrug) [inline packages/vendure-plugin-sort-by-popularity/src/sort.service.ts:63]: Typo `uniqueCollectioIds`

File: packages/vendure-plugin-sort-by-popularity/src/sort.service.ts [hunk 8]
```diff
@@ -1,12 +1,83 @@
...
+    const maxCount = groupedOrderLines[0].count;
+    const maxValue = 1000;
+    const productRepositoty =
+      this.connection.rawConnection.getRepository(Product);
+    const collectionRepositoty =
+      this.connection.rawConnection.getRepository(Collection);
+    const uniqueCollectioIds: string[] = [];
+    const collectionScoreValues: number[] = [];
+    for (const groupLines of groupedOrderLines) {
+      const score = (groupLines.count * maxValue) / maxCount;
```
Comment (martijnvdbrug) [inline packages/vendure-plugin-sort-by-popularity/src/sort.service.ts:57]: If we change the score to be an `int`, we should probably Math.round this

File: packages/vendure-plugin-sort-by-popularity/test/e2e.spec.ts [hunk 9]
```diff
@@ -84,89 +84,41 @@ describe('Sort by Popularity Plugin', function () {
...
+    );
   });
 
   it('Calculated popularity for parent collections', async () => {
-    const parentCollection = await adminClient.query<
-      Collection,
-      QueryCollectionArgs
-    >(GET_COLLECTION_ADMIN, { id: collections[0].id });
-    expect(parentCollection.customFields.popularityScore).toBe(1000);
+    expect(false).toBe(true);
```
Comment (martijnvdbrug) [inline packages/vendure-plugin-sort-by-popularity/test/e2e.spec.ts:118]: We can do this in another PR, but just to be sure we dont forget: The plugin should also calculate scores for collections that don't have products themselve, but it's subcollections do have products.

image

In this image, `category` will have the score of `subcategory.score`+`subcategory.score`

**Review by martijnvdbrug — CHANGES_REQUESTED** [review]
A few minor points, the biggest one being to use Integer instead of Float for the scores

File: packages/vendure-plugin-sort-by-popularity/src/sort.service.ts (hunk 5 above)
Comment (dalyathan) [reply packages/vendure-plugin-sort-by-popularity/src/sort.service.ts:33]: I will remove those

**Comment by martijnvdbrug** [comment]
Closing this one in favor of this one #116
